### PR TITLE
Refactor write benchmarks and split them into System Calls and Mmap files.

### PR DESCRIPTION
**Pull request by phoeinx**
This PR introduces the following changes:

- **Refactor:** Make Write Benchmarks write random values, Reuse NUMBER_OF_BYTES/ELEMENTS (which were previously calculated anew in every benchmark)

- **Extract FileIOWriteMMAPBenchmarks into its own file.**
Extracting the MMAP benchmarks into their own files allows for faster and more specific benchmark runs and makes the code leaner. This is especially relevant as we will introduce extra complexity with threaded benchmarks soon.
To not duplicate the Setup(), TearDown() and sanity_check() code it was extracted into the FileIOWriteBenchmark header. The nice way of only declaring those methods in the header and then implementing them in the corresponding cpp and including header and cpp for other build targets that need the code did not work. The benchmarks which get run by google benchmark in the file_io_write_micro_benchmark.cpp get defined unbounded and including the cpp in any build target will automatically run those benchmarks, too. Therefore to be able to have the implementation of SetUp() and co without having to run unwanted benchmarks, the implementation of said methods was put into the header file. This isn't pretty but as we are sure the header will only be included in our dedicated benchmark files I think it's acceptable.

### CODE REVIEW

File: src/benchmark/file_io_write_micro_benchmark.hpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+namespace hyrise {
+
+const auto MB = uint32_t{1'000'000};
+
+class FileIOWriteMicroBenchmarkFixture : public MicroBenchmarkBasicFixture {
+public:
+  void SetUp(::benchmark::State& state) override {
+    NUMBER_OF_BYTES = state.range(0) * MB;
+    NUMBER_OF_ELEMENTS = NUMBER_OF_BYTES / sizeof(uint32_t);
+    data_to_write = generate_random_indexes(NUMBER_OF_ELEMENTS);
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.hpp:17]: Pls rename this function

Comment (SanJSp) [reply src/benchmark/file_io_write_micro_benchmark.hpp:17]: `generate_random_data`

File: src/benchmark/file_io_write_micro_benchmark.hpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    control_sum = std::accumulate(data_to_write.begin(), data_to_write.end(), uint64_t{0});
+
+    if (creat(filename, O_RDWR) < 1) {
+      Fail("Create error:" + std::strerror(errno));
+    }
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.hpp:22]: sth like this
```suggestion
    Assert(creat(filename, O_RDWR) < 1, Fail(Create error:" + std::strerror(errno)));
    }
```

File: src/benchmark/file_io_write_micro_benchmark.hpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  const auto file_size = lseek(fd, 0, SEEK_END);
+  Assert(file_size == NUMBER_OF_BYTES, "Sanity check failed: Actual size of " + std::to_string(file_size) +
+                                       " does not match expected file size of " + std::to_string(NUMBER_OF_BYTES) +
+                                       ".");
+
+  auto read_data = std::vector<uint32_t>(NUMBER_OF_ELEMENTS);
+
+  const off_t OFFSET = 0;
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.hpp:54]:
```suggestion
  const auto OFFSET = off_t{0};
```

File: src/benchmark/file_io_write_micro_benchmark.hpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  auto* map = reinterpret_cast<uint32_t*>(mmap(NULL, NUMBER_OF_BYTES, PROT_READ, MAP_PRIVATE, fd, OFFSET));
+  close(fd);
+
+  Assert(map != MAP_FAILED, "Mapping for Sanity Check Failed:" + std::strerror(errno));
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.hpp:58]: Naming here makes no sense
```suggestion
  Assert(map != MAP_SUCCESSFULL, "Mapping for Sanity Check Failed:" + std::strerror(errno));
```

File: src/benchmark/file_io_write_mmap_benchmark.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+ *                  (1)   Random
+ *      file_size: Size argument of benchmark.
+*/
+void FileIOWriteMmapBenchmarkFixture::mmap_write_benchmark(benchmark::State& state, const int flag,
+                                                            const int data_access_mode, const ssize_t file_size) {
+  auto fd = int32_t{};
+  if ((fd = open(filename, O_RDWR)) < 0) {
+    close(fd);
+    Fail("Open error:" + std::strerror(errno));
+  }
```
Comment (SanJSp) [inline src/benchmark/file_io_write_mmap_benchmark.cpp:36]: In the future we might refactor this to an Assert with a function call? f.e.
```suggestion
  Assert((fd = open(filename, O_RDWR)) < 0,  my_fail(fd, errno));
```

File: src/benchmark/file_io_write_mmap_benchmark.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+  for (auto _ : state) {
+    state.PauseTiming();
+    micro_benchmark_clear_disk_cache();
+    state.ResumeTiming();
+
+    // Getting the mapping to memory.
+    const auto OFFSET = off_t{0};
+    int32_t* map = reinterpret_cast<int32_t*>(mmap(NULL, NUMBER_OF_BYTES, PROT_WRITE, flag, fd, OFFSET));
```
Comment (SanJSp) [inline src/benchmark/file_io_write_mmap_benchmark.cpp:51]:
```suggestion
    auto* map = reinterpret_cast<int32_t*>(mmap(NULL, NUMBER_OF_BYTES, PROT_WRITE, flag, fd, OFFSET));
```

File: src/benchmark/file_io_write_mmap_benchmark.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    switch (data_access_mode) {
+      case 0:
+        memcpy(map, std::data(data_to_write), NUMBER_OF_BYTES);
+        break;
+      case 1:
+        state.PauseTiming();
+        // Generating random indexes should not play a role in the benchmark.
+        const auto ind_access_order = generate_random_indexes(NUMBER_OF_ELEMENTS);
+        state.ResumeTiming();
+        for (uint32_t idx = 0; idx < ind_access_order.size(); ++idx) {
```
Comment (SanJSp) [inline src/benchmark/file_io_write_mmap_benchmark.cpp:63]:
```suggestion
        for (auto idx = uint32_t{0}; idx < ind_access_order.size(); ++idx) {
```

**Review by SanJSp — CHANGES_REQUESTED** [review]
Added some comments. Thank you for your work! :champagne: :smile: 
Please, lets also tackle 
- refactor mmap parameters to have "speaking" variable named flags

Possible future todo:
- refactor into super class and let the others inherit what has been moved to the hpp

File: src/benchmark/file_io_write_micro_benchmark.hpp (hunk 1 above)
Comment (phoeinx) [reply src/benchmark/file_io_write_micro_benchmark.hpp:17]: 👍 but would prefer `generate_random_numbers`

**Comment by phoeinx** [comment]
> * refactor mmap parameters to have "speaking" variable named flags
- that sounds great

> Possible future todo:
> * refactor into super class and let the others inherit what has been moved to the hpp

agreed, that would be nicer, but not sure if worth the effort / file structure complexity

File: src/benchmark/file_io_write_micro_benchmark.hpp (hunk 4 above)
Comment (phoeinx) [reply src/benchmark/file_io_write_micro_benchmark.hpp:58]: good catch!
perhaps even `map == MAP_FAILED` would work here

File: src/benchmark/file_io_write_mmap_benchmark.cpp (hunk 5 above)
Comment (phoeinx) [reply src/benchmark/file_io_write_mmap_benchmark.cpp:36]: something like `fail_and_close_file(..)` perhaps? would be prettier

**Comment by SanJSp** [comment]
Accidentally merge committed and now the history is gone :( sorry. Tried to revert it but was unable to

File: src/benchmark/file_io_write_micro_benchmark.hpp (hunk 4 above)
Comment (SanJSp) [reply src/benchmark/file_io_write_micro_benchmark.hpp:62]: `Assert` always asserts, that the given condition is true. Else the error will be printed. Hence, the naming was okay, but still feels weird.

File: src/benchmark/file_io_write_mmap_benchmark.cpp (hunk 5 above)
Comment (SanJSp) [reply src/benchmark/file_io_write_mmap_benchmark.cpp]: has been done

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 8]
```diff
@@ -92,156 +45,173 @@ BENCHMARK_DEFINE_F(FileIOWriteMicroBenchmarkFixture, WRITE_NON_ATOMIC)(benchmark
...
 
-BENCHMARK_DEFINE_F(FileIOWriteMicroBenchmarkFixture, PWRITE_ATOMIC)(benchmark::State& state) {
+void FileIOWriteMicroBenchmarkFixture::write_non_atomic_multi_threaded(benchmark::State& state, uint16_t thread_count) {
+  auto filedescriptors = std::vector<int32_t>(thread_count);
+  for (auto i = size_t{0}; i < thread_count; i++) {
+    auto fd = int32_t{};
+    if ((fd = open(filename, O_WRONLY)) < 0) {
+      close(fd);
+      Fail("Open error:" + std::strerror(errno));
+    }
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:62]: Should be assert

**Review by SanJSp — CHANGES_REQUESTED** [review]
found another thing

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 9]
```diff
@@ -7,7 +7,26 @@
...
+    Fail("Write error:" + std::strerror(errno));
+  }
+}
+
+void write_data_using_pwrite(const size_t from, const size_t to, int32_t fd, uint32_t* data_to_write_start) {
+  const auto uint32_t_size = ssize_t{sizeof(uint32_t)};
+  const auto bytes_to_write = static_cast<ssize_t>(uint32_t_size * (to - from));
+  if (pwrite(fd, data_to_write_start + from, bytes_to_write, from * uint32_t_size) != bytes_to_write) {
+    close(fd);
+    Fail("Write error:" + std::strerror(errno));
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:25]: Assert

**Review by SanJSp — CHANGES_REQUESTED** [review]
one more

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 10]
```diff
@@ -1,84 +1,37 @@
...
-  const auto sum = std::accumulate(read_data.begin(), read_data.end(), uint64_t{0});
-  Assert(control_sum == sum,
-         "Sanity check failed. Got: " + std::to_string(sum) + " Expected: " + std::to_string(control_sum));
-
-  // Remove memory mapping after job is done.
-  Assert(munmap(map, NUMBER_OF_BYTES) == 0, "Unmapping for Sanity Check failed: " + std::strerror(errno));
 }
 
-BENCHMARK_DEFINE_F(FileIOWriteMicroBenchmarkFixture, WRITE_NON_ATOMIC)(benchmark::State& state) {
+void FileIOWriteMicroBenchmarkFixture::write_non_atomic_single_threaded(benchmark::State& state) {
```
Comment (janlemcke) [inline src/benchmark/file_io_write_micro_benchmark.cpp:29]: We can use the new method 'write_data_using_write' here for consistency and to reduce code duplication.

Comment (janlemcke) [reply src/benchmark/file_io_write_micro_benchmark.cpp:29]: yes exactly. I mean it's a suggestion, i'm fine with both alternatives.

Comment (SanJSp) [reply src/benchmark/file_io_write_micro_benchmark.cpp:25]: skipped for now, future todo - this way its consistent with write

**Review by SanJSp — APPROVED** [review]
Looks great!

**Review by SanJSp — APPROVED** [review]
Yep, I like my code

**Pull request merged**